# 01. Importing Libraries

In [32]:
# Import Libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import matplotlib
import seaborn as sns
import scipy

# 02. Importing Data

In [1]:
# Project folder

path = r'D:\Data Analysis of Rental Prices in Berlin Using Python and Tableau'

In [2]:
path

'D:\\The Berlin Rental Market – An Analysis of Rental Prices in Berlin'

In [35]:
# Importing AirBnB Berlin rental data

df_berlin = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'AirBnB Berlin.csv'))


# 03. Wrangling Data

In [36]:
# Checking df shape

df_berlin.shape

(16962, 18)

## Data Types

In [37]:
df_berlin.dtypes

id                                float64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group                object
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
license                            object
dtype: object

In [38]:
df_berlin.head(5)

id                                               name    host_id  \
0  4.021868e+07  Ap 4 - Schönes Apartment nahe dem Helios Klinikum   39450649   
1  3.784921e+07          MODERN APARTMENT AT THE BERLIN WALL TRAIL  286623313   
2  4.021881e+07    Ap 33 - Schöne Ferienwohnung am Helios Klinikum   39450649   
3  3.243150e+07  Marcella`s Ferienhaus  westl. Stadtrand zu Berlin  243598169   
4  5.468740e+17  Lake house getaway in nature reserve close to Bln  156670608   

  host_name     neighbourhood_group       neighbourhood  latitude  longitude  \
0    Daniel                  Pankow                Buch  52.63136   13.51879   
1   Carsten  Tempelhof - Schöneberg         Lichtenrade  52.37272   13.41607   
2    Daniel                  Pankow                Buch  52.63136   13.51879   
3  Marcella                 Spandau  Falkenhagener Feld  52.55720   13.14205   
4      Kyra                 Spandau      Gatow / Kladow  52.45472   13.10890   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt     51               1                 11    3/4/2022   
1  Entire home/apt     40              28                  2   12/2/2019   
2  Entire home/apt     87               1                  3   11/1/2021   
3  Entire home/apt     55               3                 16    1/6/2022   
4  Entire home/apt     90               3                  4   2/27/2022   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.40                              21               345   
1               0.07                               1               352   
2               0.16                              21               304   
3               0.43                               1               215   
4               2.55                               1               251   

   number_of_reviews_ltm license  
0                      8     NaN  
1                      0     NaN  
2                      2     NaN  
3                      5     NaN  
4                      4     NaN

Based on the intitial loading of the dataset, such as the shape and head, we can already make several observations. Firstly, there are 18 columns and close to 17,000 observations. There are a few missing values, as well as many NaN observations under the license variable. With the data we have, once cleaned, we may be able to better understanding the changing AirBnB in Berlin over the last several years as well as potentially create a predictive analysis to determine what the future market may look like.

In [39]:
# Checking all Columns

pd.options.display.max_columns = None
df_berlin.head()

id                                               name    host_id  \
0  4.021868e+07  Ap 4 - Schönes Apartment nahe dem Helios Klinikum   39450649   
1  3.784921e+07          MODERN APARTMENT AT THE BERLIN WALL TRAIL  286623313   
2  4.021881e+07    Ap 33 - Schöne Ferienwohnung am Helios Klinikum   39450649   
3  3.243150e+07  Marcella`s Ferienhaus  westl. Stadtrand zu Berlin  243598169   
4  5.468740e+17  Lake house getaway in nature reserve close to Bln  156670608   

  host_name     neighbourhood_group       neighbourhood  latitude  longitude  \
0    Daniel                  Pankow                Buch  52.63136   13.51879   
1   Carsten  Tempelhof - Schöneberg         Lichtenrade  52.37272   13.41607   
2    Daniel                  Pankow                Buch  52.63136   13.51879   
3  Marcella                 Spandau  Falkenhagener Feld  52.55720   13.14205   
4      Kyra                 Spandau      Gatow / Kladow  52.45472   13.10890   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt     51               1                 11    3/4/2022   
1  Entire home/apt     40              28                  2   12/2/2019   
2  Entire home/apt     87               1                  3   11/1/2021   
3  Entire home/apt     55               3                 16    1/6/2022   
4  Entire home/apt     90               3                  4   2/27/2022   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.40                              21               345   
1               0.07                               1               352   
2               0.16                              21               304   
3               0.43                               1               215   
4               2.55                               1               251   

   number_of_reviews_ltm license  
0                      8     NaN  
1                      0     NaN  
2                      2     NaN  
3                      5     NaN  
4                      4     NaN

Observing the dataset shows us that there are several NaN values that need to be addressed before our analysis can continue

Determine which columns have null values and then using the 'sum' function will show us which variables have nulls in each column within the dataset.

In [40]:
df_berlin.isnull().sum()

id                                    0
name                                 28
host_id                               0
host_name                            14
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                        3137
reviews_per_month                  3137
calculated_host_listings_count        0
availability_365                      0
number_of_reviews_ltm                 0
license                           12367
dtype: int64

Based on these results, there are several missing observations in 'name' and 'host_name' but due to PII data concerns, these will be removed from the dataset, while the id number as well as lat/long will still provide enough information to understand the user and the general location of the apartment. 

Moreover, we can see that last_review and reviews_per_month are both missing 3,137 observations. For "last_review", is there are no reviews for the listing - the date for this listing simply does not exist. As such, the column is irrelevant and therefore amending these values is not necessary. For the variable "reviews_per_month", we can amend the missing values as 0.0, which will indicate that that the "number_of_review" column will also be 0. As such, the "reviews_per_month" would then also equal 0.0 reviews per month. 

The variable "license," refers to whether a host_id has a specific license registered regarding the renting of a room or apartment. While this may be relevant for understanding how many users actually have an official license to officially rent out an apartment, which is, AFAIK, required by German and/or Berlin law, this is not relevant for our understanding of the changing prices and growing areas of rent for our dataset. As such, this variable will likewise be removed. 

Based on the above understanding of the dataset we have at hand, we can drop several columns to ensure privacy while also removing/appending variables as necessary. 

# 04. Data Cleaning

## Dropping Columns/Variables

In [41]:
# Dropping columns that are either for this analysis irrelavant or more problematic, contain PII data.

df_berlin.drop(['name', 'host_name', 'last_review', 'license'], axis=1, inplace=True)

df_berlin.head(5)

id    host_id     neighbourhood_group       neighbourhood  \
0  4.021868e+07   39450649                  Pankow                Buch   
1  3.784921e+07  286623313  Tempelhof - Schöneberg         Lichtenrade   
2  4.021881e+07   39450649                  Pankow                Buch   
3  3.243150e+07  243598169                 Spandau  Falkenhagener Feld   
4  5.468740e+17  156670608                 Spandau      Gatow / Kladow   

   latitude  longitude        room_type  price  minimum_nights  \
0  52.63136   13.51879  Entire home/apt     51               1   
1  52.37272   13.41607  Entire home/apt     40              28   
2  52.63136   13.51879  Entire home/apt     87               1   
3  52.55720   13.14205  Entire home/apt     55               3   
4  52.45472   13.10890  Entire home/apt     90               3   

   number_of_reviews  reviews_per_month  calculated_host_listings_count  \
0                 11               0.40                              21   
1                  2               0.07                               1   
2                  3               0.16                              21   
3                 16               0.43                               1   
4                  4               2.55                               1   

   availability_365  number_of_reviews_ltm  
0               345                      8  
1               352                      0  
2               304                      2  
3               215                      5  
4               251                      4

## Replacing missing values

In [42]:
# Replacing all NaN values under "reviews_per_month" with 0
df_berlin.fillna({'reviews_per_month':0}, inplace=True)

df_berlin.reviews_per_month.isnull().sum()

0

In [43]:
# Examining changes made

df_berlin.isnull().sum()

id                                0
host_id                           0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
number_of_reviews_ltm             0
dtype: int64

In [44]:
# Examining changes made
df_berlin.shape

(16962, 14)

Based on the above changes, we no longer have any missing observations in the dataset, there are 14 distinct variables and 16,962 different observations (rows)

In [45]:
# Checking for duplicates

df_berlin.duplicated().sum()

0

# 05. Descriptive Statistics

In [46]:
# Exploring some categorical variables within the dataset

df_berlin.neighbourhood_group.unique()

array(['Pankow', 'Tempelhof - Schöneberg', 'Spandau',
       'Treptow - Köpenick', 'Steglitz - Zehlendorf', 'Neukölln',
       'Lichtenberg', 'Marzahn - Hellersdorf', 'Reinickendorf',
       'Friedrichshain-Kreuzberg', 'Mitte', 'Charlottenburg-Wilm.'],
      dtype=object)

In [47]:
df_berlin.neighbourhood.unique()

array(['Buch', 'Lichtenrade', 'Falkenhagener Feld', 'Gatow / Kladow',
       'Rahnsdorf/Hessenwinkel', 'Teltower Damm',
       'Schmöckwitz/Karolinenhof/Rauchfangswerder', 'Zehlendorf  Südwest',
       'Ostpreußendamm', 'Buckow', 'Malchow, Wartenberg und Falkenberg',
       'Altglienicke', 'Drakestr.', 'Mahlsdorf', 'West 2',
       'Prenzlauer Berg Südwest', 'Schöneberg-Nord', 'Helmholtzplatz',
       'Prenzlauer Berg Nordwest', 'nördliche Luisenstadt',
       'Reuterstraße', 'Brunnenstr. Süd', 'Tempelhofer Vorstadt',
       'Prenzlauer Berg Süd', 'Halensee', 'Regierungsviertel',
       'Frankfurter Allee Süd FK', 'Alt  Treptow', 'Neue Kantstraße',
       'Brunnenstr. Nord', 'Prenzlauer Berg Nord', 'Kantstraße',
       'Schmargendorf', 'Alexanderplatz',
       'Blankenfelde/Niederschönhausen', 'südliche Luisenstadt',
       'Schillerpromenade', 'Frankfurter Allee Nord', 'Schöneberg-Süd',
       'Südliche Friedrichstadt', 'Moabit West', 'Wiesbadener Straße',
       'Blankenburg/Heinersd

The specific neighourhood variable is very large and each is a subset of the neighbourhood_group variable. I will now examine the unique (numerical) values within the neighbourhood. 



In [48]:
len(df_berlin.neighbourhood.unique())

137

There are 137 unique neighbourhoods within Berlin, with 12 unique neighbourhood groups. 

In [49]:
# We can also quickly examine who are the top 10 hosts and what are the top 10 prices within the dataset. 

In [50]:
#The top 10 host_id listings on the AirBnB platform

top_10_host=df_berlin.host_id.value_counts().head(10)

In [51]:
top_10_host

335573202    50
8250486      49
210700814    44
378071271    38
8912086      38
108025548    30
386848467    30
3003519      30
23904144     25
1625771      25
Name: host_id, dtype: int64

In [52]:
#Based on the above, we see that several hosts own around 50 apartment units for rent. 

In [53]:
# The top 10 price of AirBnB rentals per night

top_10_price=df_berlin.price.value_counts().head(10)

In [54]:
top_10_price

50    787
40    707
30    695
35    650
60    586
45    582
25    569
80    425
55    402
70    386
Name: price, dtype: int64

In [55]:
# Based on the above we can see that their are some very expensive rental units per night, 
## which is likely not an accurate display of Berlin's AirBnB rental scene.


In [56]:
price_mean=df_berlin.price.mean()

In [57]:
price_mean

77.46474472349959

In [58]:
## Based on the above, we can see that the average price for an AirBnB rental is approximately 77.46 per night

While the above are simply only quick overviews of the dataset provided, there is certainly enough information that will allow for a proper hypothesis and analysis of the data to better understand where the most popular neighbourhoods are for renting and what the average price for that area would be. 

In [59]:
## Will finish off the initial data sourcing steps by creating a brief overview of the descriptive stats of the dataset

In [60]:
df_berlin.describe()

id       host_id      latitude     longitude         price  \
count  1.696200e+04  1.696200e+04  16962.000000  16962.000000  16962.000000   
mean   9.120904e+15  9.720243e+07     52.509949     13.404871     77.464745   
std    7.108741e+16  1.141714e+08      0.033403      0.065632    146.432360   
min    3.176000e+03  1.581000e+03     52.340070     13.082930      0.000000   
25%    1.232573e+07  1.228578e+07     52.489600     13.367660     36.000000   
50%    2.406350e+07  4.413910e+07     52.509850     13.414560     56.000000   
75%    4.023291e+07  1.484901e+08     52.533140     13.439200     89.000000   
max    5.855340e+17  4.500200e+08     52.656110     13.758550  10999.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    16962.000000       16962.000000       16962.000000   
mean        10.857859          24.325080           0.563551   
std         36.426390          55.462283           1.416329   
min          1.000000           0.000000           0.000000   
25%          2.000000           1.000000           0.030000   
50%          3.000000           5.000000           0.140000   
75%          5.000000          19.000000           0.580000   
max       1124.000000        1096.000000          91.850000   

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
count                    16962.000000      16962.000000           16962.000000  
mean                         2.910270         77.705990               4.357564  
std                          6.139532        116.864321              16.398155  
min                          1.000000          0.000000               0.000000  
25%                          1.000000          0.000000               0.000000  
50%                          1.000000          0.000000               0.000000  
75%                          2.000000        141.000000               3.000000  
max                         50.000000        365.000000             992.000000

In [61]:
print(df_berlin.describe())

                 id       host_id      latitude     longitude         price  \
count  1.696200e+04  1.696200e+04  16962.000000  16962.000000  16962.000000   
mean   9.120904e+15  9.720243e+07     52.509949     13.404871     77.464745   
std    7.108741e+16  1.141714e+08      0.033403      0.065632    146.432360   
min    3.176000e+03  1.581000e+03     52.340070     13.082930      0.000000   
25%    1.232573e+07  1.228578e+07     52.489600     13.367660     36.000000   
50%    2.406350e+07  4.413910e+07     52.509850     13.414560     56.000000   
75%    4.023291e+07  1.484901e+08     52.533140     13.439200     89.000000   
max    5.855340e+17  4.500200e+08     52.656110     13.758550  10999.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    16962.000000       16962.000000       16962.000000   
mean        10.857859          24.325080           0.563551   
std         36.426390          55.462283           1.416329   
min          1.000000           0.00

# 06. Exporting Data

In [62]:
# Exporting dataset as csv file

df_berlin.to_csv(os.path.join(path, "Data", 'Prepared Data', 'AirBnB Berlin_clean.csv'))